File: js/096_UniqueBinarySearchTrees.js
/**
 * @param {number} n
 * @return {number}
 */
var numTrees = function(n) {
    var dp = [];
    for(var k = 0; k <= n; ++k){
        dp.push(0);
    }
    dp[0] = 1;
    for(var i = 1; i <= n; ++i){
        for(var j = 0; j < i; ++j){
            dp[i] += dp[j] * dp[i-j-1];
        }
    }
    return dp[n];
};

File: cpp/096_UniqueBinarySearchTrees.cpp
class Solution {
public:
    int numTrees(int n) {
       int dp[2] = {1,2};
       for(int i = 3; i <= n; ++i){
           dp[i % 3] = dp[(i-1) % 3] * 2 + dp[(i-2) % 3];
       }
       return dp[n % 3];
    }
};